File: leetcode_ly/java/Array/WordSearch.java
public class Solution {
    public boolean exist(char[][] board, String word) {
        if(word.length() == 0) return false;
        int m = board.length;
        if(m == 0) return false;
        int n = board[0].length;
        
        boolean[][] visited = new boolean[m][n];
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                visited[i][j] = false;
            }
        }
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == word.charAt(0)){
                    visited[i][j] = true;
                    if(word.length() == 1 || search(board,word.substring(1),i,j,visited)) return true;
                    visited[i][j] = false;
                }
            }
        }
        return false;
    }
    
    public boolean search(char[][] board,String word,int i,int j,boolean[][] visited){
        int[][] direction = new int[4][2];
        direction[0][0] = direction[1][0] = direction[2][1] = direction[3][1] = 0;
        direction[0][1] = direction[2][0] = 1;
        direction[1][1] = direction[3][0] = -1;
        for(int p = 0; p < 4; p++){
            int ii = i + direction[p][0];
            int jj = j + direction[p][1];
            if(ii >= 0 && ii < board.length && jj >= 0 && jj < board[0].length && !visited[ii][jj] && board[ii][jj] == word.charAt(0)){
                visited[ii][jj] = true;
                if(word.length() == 1 || search(board,word.substring(1),ii,jj,visited)) return true;
                visited[ii][jj] = false;
            }
        }
        return false;
    }
}

File: leetcode_ly/c++/BackTracking/WordSearch.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if(word.size() == 0) return false;
        int m = board.size();
        int n = board[0].size();
        //记录是否已经走过
        vector<vector<bool>> visited(m,vector<bool>(n));
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                visited[i][j] = false;
            }
        }
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == word[0]){
                    visited[i][j] = true;
                    if(word.size() == 1 || search(board,word.substr(1),i,j,visited)) return true;
                    visited[i][j] = false;
                }
            }
        }
        return false;
    }
    
    bool search(vector<vector<char>>& board,string word,int i,int j,vector<vector<bool>>& visited){
        if(word.size() == 0) return true;
        
        //四个方向
        vector<vector<int>> direct(4,vector<int>(2));
        direct[0][0] = direct[2][1] = direct[1][0] = direct[3][1] = 0;
        direct[0][1] = direct[2][0] = 1;
        direct[3][0] = direct[1][1] = -1;
        for(int k = 0; k < direct.size(); k++){
            int ii = i + direct[k][0];
            int jj = j + direct[k][1];
            if(ii >= 0 && jj >= 0 && ii < board.size() && jj < board[0].size() && !visited[ii][jj] && board[ii][jj] == word[0]){
                visited[ii][jj] = true;
                if(word.size() == 1 || search(board,word.substr(1),ii,jj,visited)) return true;
                visited[ii][jj] = false;
            }
        }
        return false;
    }
};